x-pack/plugins/alerts/server/invalidate_pending_api_keys/task.ts

```diff
@@ -0,0 +1,195 @@
+  coreStartServices: Promise<[CoreStart, AlertingPluginsStart, unknown]>,
+  taskManager: TaskManagerSetupContract,
+  config: Promise<AlertsConfig>,
+  securityPluginSetup?: SecurityPluginSetup
+) {
+  taskManager.registerTaskDefinitions({
+    [TASK_TYPE]: {
+      title: 'Invalidate alert API Keys',
+      createTaskRunner: taskRunner(logger, coreStartServices, config, securityPluginSetup),
+    },
+  });
+}
+
+function taskRunner(
+  logger: Logger,
+  coreStartServices: Promise<[CoreStart, AlertingPluginsStart, unknown]>,
+  config: Promise<AlertsConfig>,
+  securityPluginSetup?: SecurityPluginSetup
+) {
+  return ({ taskInstance }: RunContext) => {
+    const { state } = taskInstance;
+    return {
+      async run() {
+        let totalInvalidated = 0;
+        const configResult = await config;
+        try {
+          const [{ savedObjects }] = await coreStartServices;
+          const repository = savedObjects.createInternalRepository([
+            'api_key_pending_invalidation',
+          ]);
+          const configuredDelay = configResult.invalidateApiKeysTask.removalDelay;
+          const delay = timePeriodBeforeDate(new Date(), configuredDelay).toISOString();
+
+          let hasApiKeysPendingInvalidation = true;
+          const PAGE_SIZE = 100;
+          do {
+            const apiKeysToInvalidate = await repository.find<InvalidatePendingApiKey>({
+              type: 'api_key_pending_invalidation',
+              filter: `api_key_pending_invalidation.attributes.createdAt <= "${delay}"`,
+              page: 1,
+              sortField: 'createdAt',
+              sortOrder: 'asc',
+              perPage: PAGE_SIZE,
+            });
+            totalInvalidated += await invalidateApiKeys(
+              logger,
+              repository,
+              apiKeysToInvalidate,
+              securityPluginSetup
+            );
+
+            hasApiKeysPendingInvalidation = apiKeysToInvalidate.total > PAGE_SIZE;
+          } while (hasApiKeysPendingInvalidation);
+
+          return {
+            state: {
+              runs: (state.runs || 0) + 1,
+              total_invalidated: totalInvalidated,
+            },
+            schedule: {
+              interval: configResult.invalidateApiKeysTask.interval,
+            },
+          };
+        } catch (e) {
+          logger.warn(`Error executing alerting apiKey invalidation task: ${e.message}`);
+          return {
+            state: {
+              runs: (state.runs || 0) + 1,
+              total_invalidated: totalInvalidated,
+            },
+            schedule: {
+              interval: configResult.invalidateApiKeysTask.interval,
+            },
+          };
+        }
+      },
+    };
+  };
+}
+
+async function invalidateApiKeys(
+  logger: Logger,
+  repository: ISavedObjectsRepository,
+  apiKeysToInvalidate: SavedObjectsFindResponse<InvalidatePendingApiKey>,
+  securityPluginSetup?: SecurityPluginSetup
+) {
+  let totalInvalidated = 0;
+  await Promise.all(
```

Comment:
So this means we sending all the invalidation requests at once?  I'm worried about overwhelming ES here - should be breaking this into something like 10 or 20 to run at once, as a set of batches?

We should also ask security about a batch version of invalidating these.

Reply:
I think we should have a debug log call here.  Probably just one for all the keys, just printing `totalInvalidated`, at the end.

Reply:
Security team has an opened issue opened for the batch invalidation by Ids array https://github.com/elastic/kibana/issues/79714, so I will open a follow up issue on their's with the proper dependency on the issue

Reply:
Isn't it currently restricted by the page size of 100 in the `const apiKeysToInvalidate = await savedObjectsClient.find<InvalidatePendingApiKey>`? I think having 100 maximum won't impact Kibana so hard.

Reply:
I may reduce the number in twice - make it 50 :-)